File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.cs
using System.Collections.Generic;

public class Solution {
    public string[] solution(string[] players, string[] callings) {
        string[] answer = new string[players.Length];
        Dictionary<string, int> pd = new Dictionary<string, int>();
        Dictionary<int, string> id = new Dictionary<int, string>();
        
        for (int i = 0; i < players.Length; i++)
        {
            pd.Add(players[i], i);
            id.Add(i, players[i]);
        }
        
        foreach (string nowPlayer in callings)
        {
            int nowRank = pd[nowPlayer];
            string frontPlayer = id[nowRank - 1];
            
            pd[nowPlayer] -= 1;
            pd[frontPlayer] += 1;
            
            id[nowRank - 1] = nowPlayer;
            id[nowRank] = frontPlayer;
        }
        
        for (int i = 0; i < players.Length; i++)
        {
            answer[i] = id[i];
        }
        
        return answer;
    }
}

File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    pd = {v:i for i,v in enumerate(players)}
    rd = {i:v for i,v in enumerate(players)}
    
    for now in callings:
        nowRank = pd[now]
        front = rd[nowRank - 1]
        
        pd[now], pd[front], rd[nowRank], rd[nowRank - 1] = pd[front], pd[now], rd[nowRank-1], rd[nowRank]
        
    return [rd[i] for i in range(len(players))]